plugins/inputs/elasticsearch/elasticsearch.go

```diff
@@ -153,50 +115,55 @@ func (e *Elasticsearch) Gather(acc telegraf.Accumulator) error {
 		e.client = client
 	}
 
-	errChan := errchan.New(len(e.Servers) * 3)
+	errChan := errchan.New(len(e.Servers))
 	var wg sync.WaitGroup
 	wg.Add(len(e.Servers))
 
+	needCatMaster := e.ClusterStats || e.IndicesStats || e.ShardsStats
+
 	for _, serv := range e.Servers {
 		go func(s string, acc telegraf.Accumulator) {
 			defer wg.Done()
+
+			// Gather cat master when
+			e.localNodeIsMaster = false
+			if needCatMaster {
+				if err := e.gatherCatMaster(s+"/_cat/master", acc); err != nil {
+					errChan.C <- err
+					return
+				}
+			}
+
+			// Always gather node stats
 			var url string
 			if e.Local {
 				url = s + statsPathLocal
 			} else {
 				url = s + statsPath
 			}
-			e.isMaster = false
-
-			if e.ClusterStats {
-				// get cat/master information here so NodeStats can determine
-				// whether this node is the Master
-				e.setCatMaster(s + "/_cat/master")
-			}
-
-			// Always gather node states
 			if err := e.gatherNodeStats(url, acc); err != nil {
-				err = fmt.Errorf(mask.ReplaceAllString(err.Error(), "http(s)://XXX:XXX@"))
 				errChan.C <- err
 				return
 			}
 
+			// Optional stats
+
 			if e.ClusterHealth {
-				url = s + "/_cluster/health?level=indices"
-				if err := e.gatherClusterHealth(url, acc); err != nil {
-					err = fmt.Errorf(mask.ReplaceAllString(err.Error(), "http(s)://XXX:XXX@"))
-					errChan.C <- err
-					return
-				}
+				e.gatherClusterHealth(s+"/_cluster/health?level=indices", acc)
 			}
 
-			if e.ClusterStats && e.isMaster {
-				if err := e.gatherClusterStats(s+"/_cluster/stats", acc); err != nil {
-					err = fmt.Errorf(mask.ReplaceAllString(err.Error(), "http(s)://XXX:XXX@"))
-					errChan.C <- err
-					return
-				}
+			if e.ClusterStats && e.localNodeIsMaster {
+				e.gatherClusterStats(s+"/_cluster/stats", acc)
+			}
+
+			if e.IndicesStats && !e.ShardsStats && e.localNodeIsMaster {
+				e.gatherIndicesStats(s+"_all/_stats", acc, false)
```

Comment:
It seems reasonable to think that /_stats is the same as /_all/_stats. The _all part just specifies all indices for the index-name portion so it returns all index level stats (which is what I wanted).  However, /_stats indicates that it returns high level aggregation as well, so I'm not sure whether that might be different (and hence not part of the metrics that the plugin is now gathering).